bootstrap/src/sun/nio/ch/lincheck/Types.java

```diff
@@ -0,0 +1,185 @@
+        switch (className) {
+            case "I": return INT_TYPE;
+            case "J": return LONG_TYPE;
+            case "D": return DOUBLE_TYPE;
+            case "F": return FLOAT_TYPE;
+            case "Z": return BOOLEAN_TYPE;
+            case "B": return BYTE_TYPE;
+            case "S": return SHORT_TYPE;
+            case "C": return CHAR_TYPE;
+            default:
+                if (className.startsWith("[")) {
+                    return new ArrayType(convertAsmArgumentTypeName(className.substring(1)));
+                } else {
+                    if (!className.startsWith("L") || !className.endsWith(";")) {
+                        throw new IllegalArgumentException("Invalid type name: " + className);
+                    }
+                    return new ObjectType(className.substring(1, className.length() - 1).replace('/', '.'));
+                }
+        }
+    }
+
+    private static Type convertAsmTypeName(String className) {
+        if ("V".equals(className)) {
+            return VOID_TYPE;
+        } else {
+            return convertAsmArgumentTypeName(className);
+        }
+    }
+
+    /**
+     * Parses each descriptor substring of the `methodDesc`, including the return type of the method.
+     *
+     * @param methodDesc descriptor of the method.
+     */
+    public static MethodType convertAsmMethodType(String methodDesc) {
+        // Modified code for parsing the type descriptors
+        // in the method descriptor (which looks like this: "(args...)ret")
+        List<Type> argumentTypes = new ArrayList<>();
+        int currentOffset = 1;
+
+        while (methodDesc.charAt(currentOffset) != ')') {
+            int currentDescriptorTypeOffset = currentOffset;
+            while (methodDesc.charAt(currentOffset) == '[') {
+                currentOffset++;
+            }
+            if (methodDesc.charAt(currentOffset++) == 'L') {
+                int semiColumnOffset = methodDesc.indexOf(';', currentOffset);
+                currentOffset = Math.max(currentOffset, semiColumnOffset + 1);
+            }
+            argumentTypes.add(
+                    convertAsmTypeName(methodDesc.substring(currentDescriptorTypeOffset, currentOffset))
+            );
+        }
+        Type returnType = convertAsmTypeName(methodDesc.substring(currentOffset + 1));
+
+        return new MethodType(argumentTypes, returnType);
+    }
+
+    public static final VoidType VOID_TYPE = new VoidType();
+    public static final IntType INT_TYPE = new IntType();
+    public static final LongType LONG_TYPE = new LongType();
+    public static final DoubleType DOUBLE_TYPE = new DoubleType();
+    public static final FloatType FLOAT_TYPE = new FloatType();
+    public static final BooleanType BOOLEAN_TYPE = new BooleanType();
+    public static final ByteType BYTE_TYPE = new ByteType();
+    public static final ShortType SHORT_TYPE = new ShortType();
+    public static final CharType CHAR_TYPE = new CharType();
```

Comment:
Nitpicking, but I personally like to format such code patterns as follows:
```java
    public static final VoidType VOID_TYPE = new VoidType();
    public static final  IntType  INT_TYPE = new IntType();
    public static final LongType LONG_TYPE = new LongType();
    ...
```  

Makes it easier to read the code IMO.